### PR TITLE
Feature/d3 asim 3493 Adapted PV settings to enable new power scaling factor capacity_kW

### PR DESCRIPTION
**Pull request by hannesdiedrich**

### CODE REVIEW

File: d3a_interface/settings_validators.py [hunk 1]
```diff
@@ -15,75 +15,77 @@
...
+                       if isinstance(slot_length, timedelta) else duration(minutes=slot_length))
 
     if "tick_length" in settings_dict:
         min_tick_length, max_tick_length = calc_min_max_tick_length(slot_length)
         if not min_tick_length <= tick_length <= max_tick_length:
-            raise D3ASettingsException(f'Invalid tick_length '
-                                       f'({tick_length.in_seconds()} sec, limits: '
-                                       f'[{min_tick_length.in_seconds()} sec, '
-                                       f'{max_tick_length.in_seconds()} sec])')
+            raise D3ASettingsException(f"Invalid tick_length "
```
Comment (Hassan754) [inline d3a_interface/settings_validators.py:43]: nit
```suggestion
            raise D3ASettingsException("Invalid tick_length "
```

File: d3a_interface/settings_validators.py [hunk 2]
```diff
@@ -15,75 +15,77 @@
...
+                                       f"{max_tick_length.in_seconds()} sec])")
     if "slot_length" in settings_dict:
         min_slot_length, max_slot_length = calc_min_max_slot_length(tick_length)
         if not min_slot_length <= slot_length <= max_slot_length:
-            raise D3ASettingsException(f'Invalid slot_length '
-                                       f'({slot_length.in_minutes()} min, limits: '
-                                       f'[{min_slot_length.in_minutes()} min, '
-                                       f'{max_slot_length.in_minutes()} min])')
-    if "cloud_coverage" in settings_dict and \
+            raise D3ASettingsException(f"Invalid slot_length "
```
Comment (Hassan754) [inline d3a_interface/settings_validators.py:50]:
```suggestion
            raise D3ASettingsException("Invalid slot_length "
```

**Review by Hassan754 — APPROVED** [review]
LGTM

File: d3a_interface/settings_validators.py [hunk 3]
```diff
@@ -15,75 +15,77 @@
...
 
 def validate_global_settings(settings_dict):
-    slot_length = settings_dict["slot_length"] \
-        if "slot_length" in settings_dict else GlobalConfig.slot_length
-    tick_length = settings_dict["tick_length"] \
-        if "tick_length" in settings_dict else GlobalConfig.tick_length
+    slot_length = (settings_dict["slot_length"]
+                   if "slot_length" in settings_dict else GlobalConfig.slot_length)
+    tick_length = (settings_dict["tick_length"]
+                   if "tick_length" in settings_dict else GlobalConfig.tick_length)
```
Comment (faizan2590) [inline d3a_interface/settings_validators.py:30]:
```suggestion
    tick_length = getattr(settings_dict, "tick_length", GlobalConfig.tick_length)
```
Seems better to catch default values

**Review by faizan2590 — APPROVED** [review]
LGTM, though one enhancement comment

File: d3a_interface/settings_validators.py (hunk 3 above)
Comment (faizan2590) [reply d3a_interface/settings_validators.py:30]: Sorry, my suggestion wont work here as its a dict and not a python object

File: d3a_interface/settings_validators.py [hunk 4]
```diff
@@ -15,75 +15,77 @@
...
-            <= settings_dict["max_panel_power_W"] \
-            <= ConstSettings.PVSettings.MAX_PANEL_OUTPUT_W_LIMIT[1]:
-        raise D3ASettingsException(f'Invalid value for max_panel_power_W '
-                                   f'({settings_dict["max_panel_power_W"]}).')
-    if "grid_fee_type" in settings_dict and \
-            int(settings_dict["grid_fee_type"]) not in ConstSettings.IAASettings.VALID_FEE_TYPES:
-        raise D3ASettingsException(f'Invalid value for grid_fee_type '
-                                   f'({settings_dict["grid_fee_type"]}).')
+    if ("capacity_kW" in settings_dict and not
+            ConstSettings.PVSettings.CAPACITY_KW_LIMIT[0]
```
Comment (fievelk) [inline d3a_interface/settings_validators.py:72]: `RangeLimit` has `min` and `max` attributes for this purpose. They would help to make the code clearer:
```suggestion
            ConstSettings.PVSettings.CAPACITY_KW_LIMIT.min
```

Comment (fievelk) [reply d3a_interface/settings_validators.py:72]: The same applies for `CAPACITY_KW_LIMIT[1]`.

File: d3a_interface/validators/pv_validator.py [hunk 5]
```diff
@@ -43,7 +43,16 @@ def validate_energy(cls, **kwargs):
...
+        if kwargs.get("capacity_kW") is not None:
+            error_message = {
+                "misconfiguration": ["capacity_kW should be in between "
+                                     f"{PvSettings.CAPACITY_KW_LIMIT.min} & "
+                                     f"{PvSettings.CAPACITY_KW_LIMIT.max}"]}
+            validate_range_limit(PvSettings.CAPACITY_KW_LIMIT.min,
+                                 kwargs["capacity_kW"],
+                                 PvSettings.CAPACITY_KW_LIMIT.max, error_message)
+
+        if kwargs.get("max_panel_power_W") is not None:  # deprecated
```
Comment (fievelk) [inline d3a_interface/validators/pv_validator.py:55]: Thanks for adding the `# deprecated` comment!

File: tests/test_settings_validators.py [hunk 6]
```diff
@@ -74,13 +74,13 @@ def test_wrong_cloud_coverage(self):
         self.assertRaises(D3ASettingsException, validate_global_settings,
                           {"cloud_coverage": ConstSettings.PVSettings.CLOUD_COVERAGE_LIMIT[1] + 1})
 
-    def test_wrong_max_panel_power_W(self):
+    def test_wrong_capacity_kW(self):
         self.assertRaises(D3ASettingsException, validate_global_settings,
-                          {"max_panel_power_W":
-                           ConstSettings.PVSettings.MAX_PANEL_OUTPUT_W_LIMIT[0] - 1})
+                          {"capacity_kW":
+                           ConstSettings.PVSettings.CAPACITY_KW_LIMIT[0] - 1})
```
Comment (fievelk) [inline tests/test_settings_validators.py:80]:
```suggestion
                           ConstSettings.PVSettings.CAPACITY_KW_LIMIT.min - 1})
```

File: tests/test_settings_validators.py [hunk 7]
```diff
@@ -74,13 +74,13 @@ def test_wrong_cloud_coverage(self):
...
         self.assertRaises(D3ASettingsException, validate_global_settings,
-                          {"max_panel_power_W":
-                           ConstSettings.PVSettings.MAX_PANEL_OUTPUT_W_LIMIT[1] + 1})
+                          {"capacity_kW":
+                           ConstSettings.PVSettings.CAPACITY_KW_LIMIT[1] + 1})
```
Comment (fievelk) [inline tests/test_settings_validators.py:83]:
```suggestion
                           ConstSettings.PVSettings.CAPACITY_KW_LIMIT.max + 1})
```

File: tox.ini [hunk 8]
```diff
@@ -31,8 +31,8 @@ commands =
     python -c "import fcntl; fcntl.fcntl(1, fcntl.F_SETFL, 0)"
     pip-sync requirements/tests.txt
     pip install -e .
-    flake8
     coverage run -p -m pytest tests
     coverage combine
     coverage xml
     coverage report
+    flake8
```
Comment (fievelk) [inline tox.ini:38]: Why do you want to move it to the end? `flake` is faster than unit tests, so running it before them allows to stop the process immediately without having to wait for tests.

**Review by fievelk — DISMISSED** [review]
Left some comments. Nothing huge, so I'll give my LGTM.

File: tox.ini (hunk 8 above)
Comment (hannesdiedrich) [reply tox.ini:38]: I really can not remember. You are right, will change it back.

**Review by Hassan754 — APPROVED** [review]

**Review by fievelk — APPROVED** [review]

**Pull request merged**